### ISSUE TITLE
Interaction of level down args (e.g., `maxlevel`, `adjustlevel`) with random battles

### ISSUE THREAD

**Issue by DiTo97** [smogon/pokemon-showdown#9392]
Hi @Zarel, @Marty-D,

As originally pointed out in #9390, level down args (e.g., `maxlevel`, `adjustlevel`) seem not to have effect on random battles, whenever a custom battle with custom rules is defined. In my experiments, I tested the following ruling:

`/challenge gen8randomdoublesbattle @@@ adjustleveldown=50, maxteamsize=6, pickedteamsize=4, teampreview, vgctimer`

**Warning:** I have no idea of the implementation behind the generation of random teams. The following is purely speculative.

## Team generation

I can imagine two different approaches for generating random teams in random battles:
- *level-aware*, that uses the levels of 'mons to condition the generation of teams
- *non-level-aware*, that focuses on level scaling only after the generation of teams

When I say non-level-aware generation I imagine something roughly like this:
1. select the two teams of six 'mons randomly (maybe with some heuristics);
2. compute a joint _strength_ score that accounts for level and BST (it could be as simple as the product);
3. optimize the levels of 'mons with a DP algorithm that balances the total strength score between the two teams.

Adding the level down constraint to the former approach would be trivial, as the level 50 constraint would condition the generation, although likely limiting the search space of random teams that can be generated; on the other end, it would be almost impossible to enforce the level constraint to the latter approach, as it would greatly increase the risk of generating imbalanced teams.

Now, unnecessary details aside, what is approach that comes closest to the generation of random teams in random battles? In other words, how (non-)trivial would it be to add a level constraint to the generation, e.g., via `adjustleveldown`?

**Comment by adrivrie** [comment]
Random team generation happens [here](https://github.com/smogon/pokemon-showdown/blob/master/data/random-teams.ts), at least for gen 9, but the others are similar enough. At line 143 `adjustLevel` is read from the ruleset, and it's applied if available later (specifically at line 1256). Indeed, using `adjustlevel=50` works, or at least it should.

By default these extra rules only affect the team validator, so they have to be explicitly included in random team generation if you want them to have an effect on random battles. This is not impossible, but I don't think there's any real usecase that cannot be captured by `adjustlevel=`.

As for how random battles generation works: Six random mons are selected, indeed with heuristics, and each pokemon has a level associated with it. For gens 8 and 9 you can see these levels [here](https://github.com/smogon/pokemon-showdown/blob/master/data/mods/gen8/formats-data.ts) and [here](https://github.com/smogon/pokemon-showdown/blob/master/data/random-sets.json) respectively. They have been empirically derived and are not the result of a formula.

**Comment by DiTo97** [comment]
Thank you @adrivrie,

I tested with `adjustlevel` and it work. I was focusing on `adjustleveldown`, as that is what VGC actually does, and I was looking to create VGC-like random double battles, as I explained in the other issue. Anyways, it's a subtle difference that covers really narrow edge cases (the infamous Lv. 49s under TR), so I would say that `adjustlevel` is more than fine as it is.

As for the random teams generations, does the selection heuristic consider the level? I understand that it is probably for the better to have fixed-level ready-made sets for the 'mons, but if the heuristic considers the set as it is, and the level scaling with `adjustlevel` happens afterwards, that may likely turn balanced teams into imbalanced Lv. 50 teams, right?.

**Comment by adrivrie** [comment]
Indeed, using adjustlevel will make wildly imbalanced teams. It's mostly used either for the memes or in combination with another mod that balances it out (such as scalemons or even tiershift). There's currently no good way to make randomized VGC.
Adjusting random battles such that it looks more like VGC can of course be fun, but it's not balanced, nor will it really resemble VGC in playstyle.

**Comment by DiTo97** [comment]
Yeah, as I was guessing, @adrivrie.

Back to my question, how difficult would it be to run the heuristic selection algorithm on scaled levels rather than the original hand-made ones?

It would be enough if you could just point me to what changes I would have to make to apply this change which I think would surely alleviate a bit the imbalance problem. I will probably fork the repository to keep things clean and start the experimentation on VGC-like battle bots.

**Comment by adrivrie** [comment]
Anything that otherwise only affects team validation is by default ignored in randoms. `adjustlevel` is the exception in that regard.

Also, to perhaps clear up some confusion, there isn't really any selection algorithm in play. For every slot an entirely random mon is selected (from the pool of all fully evolved pokemon and a couple of NFE ones which fill an interesting niche), and is added to the team unless it fails a specific requirement. This rejection is quite rare, so it really is almost entirely random. The requirements vary slightly between gens, but generally they cap any team at 1 mon with the same typing, 2 that share a type and 3 that share a weakness. Older gens also softcap the amount of pokemon from any given smogon singles tier.

My point is that the heuristics (if we can call it that) don't take levels into account, nor do they do all that much. You could remove all of them and the teams would look roughly similar, just with a decent chance of rolling two mono-water mons.

So, if you want to make a random VGC that uses a certain algorithm for determining teams, you'll have to remake `randomTeam()` in its entirety. If the normal random battles sets are good enough, then `randomSet()` and the boilerplate of all the counters can remain, but otherwise there's no system in place to adapt.

**Comment by DiTo97** [comment]
Alright, @adrivrie.

This means that the "actual" selection algorithm is in the empirical hand-made tuning of levels and sets that you already do. This way, whatever 'mon is picked up by the team, barring those minor invalidations you mentioned, the team will be decently balanced by definition.

Is that a fair assessment?

I will have to revise `randomSet` a bit more, but I think that is good already. If I am not mistaken, it just selects a random set across all those available for a given species, barring items and maybe abilities that are added later. I will probably modify `randomTeam` to something similar to what I described in the first comment, without neglecting all the good empirical work you have done on level scaling. For instance, assuming that the more powerful a 'mon is (in terms of BST or whatever) the lower is level will be in the empirical sets, I could do the following:
- compute a *weight* score for all sets (I saw different sets of the same species may be attributed with different levels) as the product of BST and the empirical level;
- generare $n$ teams with a DP knapsack algorithm that are above an empirical minimum threshold of total weight (accounting for `speciesclause`, if set, or the type overlap heuristics you were mentioning above);
- select two teams out of the $n$;
- attribute abilities and items (accounting for `itemclause`, if set).

If the emperitical threshold is tuned properly, then is would still leave room for a ton of variety in the generation process, while retaining most of the balance of the hand-made sets.

If `adjustlevel` is not set, then the empirical threshold can be set two zero as any team composition would be okay, as the level scaling is already some empirically. This would essentially be compatibile with how it is now.

Thank a lot for the support!

**Comment by joebangara43903** [comment]
![image](https://user-images.githubusercontent.com/131228993/232964139-9c376468-2417-4f96-8df4-29531504b69e.png)
